// @vue-storefront/magento-types/lib/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

exports.CartUserInputErrorType = void 0;
(function (CartUserInputErrorType) {
    CartUserInputErrorType["InsufficientStock"] = "INSUFFICIENT_STOCK";
    CartUserInputErrorType["NotSalable"] = "NOT_SALABLE";
    CartUserInputErrorType["ProductNotFound"] = "PRODUCT_NOT_FOUND";
    CartUserInputErrorType["Undefined"] = "UNDEFINED";
})(exports.CartUserInputErrorType || (exports.CartUserInputErrorType = {}));
exports.CheckoutAgreementMode = void 0;
(function (CheckoutAgreementMode) {
    CheckoutAgreementMode["Auto"] = "AUTO";
    CheckoutAgreementMode["Manual"] = "MANUAL";
})(exports.CheckoutAgreementMode || (exports.CheckoutAgreementMode = {}));
exports.CheckoutUserInputErrorCodes = void 0;
(function (CheckoutUserInputErrorCodes) {
    CheckoutUserInputErrorCodes["InsufficientStock"] = "INSUFFICIENT_STOCK";
    CheckoutUserInputErrorCodes["NotSalable"] = "NOT_SALABLE";
    CheckoutUserInputErrorCodes["ProductNotFound"] = "PRODUCT_NOT_FOUND";
    CheckoutUserInputErrorCodes["ReorderNotAvailable"] = "REORDER_NOT_AVAILABLE";
    CheckoutUserInputErrorCodes["Undefined"] = "UNDEFINED";
})(exports.CheckoutUserInputErrorCodes || (exports.CheckoutUserInputErrorCodes = {}));
/** The list of countries codes */
exports.CountryCodeEnum = void 0;
(function (CountryCodeEnum) {
    /** Andorra */
    CountryCodeEnum["Ad"] = "AD";
    /** United Arab Emirates */
    CountryCodeEnum["Ae"] = "AE";
    /** Afghanistan */
    CountryCodeEnum["Af"] = "AF";
    /** Antigua & Barbuda */
    CountryCodeEnum["Ag"] = "AG";
    /** Anguilla */
    CountryCodeEnum["Ai"] = "AI";
    /** Albania */
    CountryCodeEnum["Al"] = "AL";
    /** Armenia */
    CountryCodeEnum["Am"] = "AM";
    /** Netherlands Antilles */
    CountryCodeEnum["An"] = "AN";
    /** Angola */
    CountryCodeEnum["Ao"] = "AO";
    /** Antarctica */
    CountryCodeEnum["Aq"] = "AQ";
    /** Argentina */
    CountryCodeEnum["Ar"] = "AR";
    /** American Samoa */
    CountryCodeEnum["As"] = "AS";
    /** Austria */
    CountryCodeEnum["At"] = "AT";
    /** Australia */
    CountryCodeEnum["Au"] = "AU";
    /** Aruba */
    CountryCodeEnum["Aw"] = "AW";
    /** Åland Islands */
    CountryCodeEnum["Ax"] = "AX";
    /** Azerbaijan */
    CountryCodeEnum["Az"] = "AZ";
    /** Bosnia & Herzegovina */
    CountryCodeEnum["Ba"] = "BA";
    /** Barbados */
    CountryCodeEnum["Bb"] = "BB";
    /** Bangladesh */
    CountryCodeEnum["Bd"] = "BD";
    /** Belgium */
    CountryCodeEnum["Be"] = "BE";
    /** Burkina Faso */
    CountryCodeEnum["Bf"] = "BF";
    /** Bulgaria */
    CountryCodeEnum["Bg"] = "BG";
    /** Bahrain */
    CountryCodeEnum["Bh"] = "BH";
    /** Burundi */
    CountryCodeEnum["Bi"] = "BI";
    /** Benin */
    CountryCodeEnum["Bj"] = "BJ";
    /** St. Barthélemy */
    CountryCodeEnum["Bl"] = "BL";
    /** Bermuda */
    CountryCodeEnum["Bm"] = "BM";
    /** Brunei */
    CountryCodeEnum["Bn"] = "BN";
    /** Bolivia */
    CountryCodeEnum["Bo"] = "BO";
    /** Brazil */
    CountryCodeEnum["Br"] = "BR";
    /** Bahamas */
    CountryCodeEnum["Bs"] = "BS";
    /** Bhutan */
    CountryCodeEnum["Bt"] = "BT";
    /** Bouvet Island */
    CountryCodeEnum["Bv"] = "BV";
    /** Botswana */
    CountryCodeEnum["Bw"] = "BW";
    /** Belarus */
    CountryCodeEnum["By"] = "BY";
    /** Belize */
    CountryCodeEnum["Bz"] = "BZ";
    /** Canada */
    CountryCodeEnum["Ca"] = "CA";
    /** Cocos (Keeling) Islands */
    CountryCodeEnum["Cc"] = "CC";
    /** Congo-Kinshasa */
    CountryCodeEnum["Cd"] = "CD";
    /** Central African Republic */
    CountryCodeEnum["Cf"] = "CF";
    /** Congo-Brazzaville */
    CountryCodeEnum["Cg"] = "CG";
    /** Switzerland */
    CountryCodeEnum["Ch"] = "CH";
    /** Côte d’Ivoire */
    CountryCodeEnum["Ci"] = "CI";
    /** Cook Islands */
    CountryCodeEnum["Ck"] = "CK";
    /** Chile */
    CountryCodeEnum["Cl"] = "CL";
    /** Cameroon */
    CountryCodeEnum["Cm"] = "CM";
    /** China */
    CountryCodeEnum["Cn"] = "CN";
    /** Colombia */
    CountryCodeEnum["Co"] = "CO";
    /** Costa Rica */
    CountryCodeEnum["Cr"] = "CR";
    /** Cuba */
    CountryCodeEnum["Cu"] = "CU";
    /** Cape Verde */
    CountryCodeEnum["Cv"] = "CV";
    /** Christmas Island */
    CountryCodeEnum["Cx"] = "CX";
    /** Cyprus */
    CountryCodeEnum["Cy"] = "CY";
    /** Czech Republic */
    CountryCodeEnum["Cz"] = "CZ";
    /** Germany */
    CountryCodeEnum["De"] = "DE";
    /** Djibouti */
    CountryCodeEnum["Dj"] = "DJ";
    /** Denmark */
    CountryCodeEnum["Dk"] = "DK";
    /** Dominica */
    CountryCodeEnum["Dm"] = "DM";
    /** Dominican Republic */
    CountryCodeEnum["Do"] = "DO";
    /** Algeria */
    CountryCodeEnum["Dz"] = "DZ";
    /** Ecuador */
    CountryCodeEnum["Ec"] = "EC";
    /** Estonia */
    CountryCodeEnum["Ee"] = "EE";
    /** Egypt */
    CountryCodeEnum["Eg"] = "EG";
    /** Western Sahara */
    CountryCodeEnum["Eh"] = "EH";
    /** Eritrea */
    CountryCodeEnum["Er"] = "ER";
    /** Spain */
    CountryCodeEnum["Es"] = "ES";
    /** Ethiopia */
    CountryCodeEnum["Et"] = "ET";
    /** Finland */
    CountryCodeEnum["Fi"] = "FI";
    /** Fiji */
    CountryCodeEnum["Fj"] = "FJ";
    /** Falkland Islands */
    CountryCodeEnum["Fk"] = "FK";
    /** Micronesia */
    CountryCodeEnum["Fm"] = "FM";
    /** Faroe Islands */
    CountryCodeEnum["Fo"] = "FO";
    /** France */
    CountryCodeEnum["Fr"] = "FR";
    /** Gabon */
    CountryCodeEnum["Ga"] = "GA";
    /** United Kingdom */
    CountryCodeEnum["Gb"] = "GB";
    /** Grenada */
    CountryCodeEnum["Gd"] = "GD";
    /** Georgia */
    CountryCodeEnum["Ge"] = "GE";
    /** French Guiana */
    CountryCodeEnum["Gf"] = "GF";
    /** Guernsey */
    CountryCodeEnum["Gg"] = "GG";
    /** Ghana */
    CountryCodeEnum["Gh"] = "GH";
    /** Gibraltar */
    CountryCodeEnum["Gi"] = "GI";
    /** Greenland */
    CountryCodeEnum["Gl"] = "GL";
    /** Gambia */
    CountryCodeEnum["Gm"] = "GM";
    /** Guinea */
    CountryCodeEnum["Gn"] = "GN";
    /** Guadeloupe */
    CountryCodeEnum["Gp"] = "GP";
    /** Equatorial Guinea */
    CountryCodeEnum["Gq"] = "GQ";
    /** Greece */
    CountryCodeEnum["Gr"] = "GR";
    /** South Georgia & South Sandwich Islands */
    CountryCodeEnum["Gs"] = "GS";
    /** Guatemala */
    CountryCodeEnum["Gt"] = "GT";
    /** Guam */
    CountryCodeEnum["Gu"] = "GU";
    /** Guinea-Bissau */
    CountryCodeEnum["Gw"] = "GW";
    /** Guyana */
    CountryCodeEnum["Gy"] = "GY";
    /** Hong Kong SAR China */
    CountryCodeEnum["Hk"] = "HK";
    /** Heard &amp; McDonald Islands */
    CountryCodeEnum["Hm"] = "HM";
    /** Honduras */
    CountryCodeEnum["Hn"] = "HN";
    /** Croatia */
    CountryCodeEnum["Hr"] = "HR";
    /** Haiti */
    CountryCodeEnum["Ht"] = "HT";
    /** Hungary */
    CountryCodeEnum["Hu"] = "HU";
    /** Indonesia */
    CountryCodeEnum["Id"] = "ID";
    /** Ireland */
    CountryCodeEnum["Ie"] = "IE";
    /** Israel */
    CountryCodeEnum["Il"] = "IL";
    /** Isle of Man */
    CountryCodeEnum["Im"] = "IM";
    /** India */
    CountryCodeEnum["In"] = "IN";
    /** British Indian Ocean Territory */
    CountryCodeEnum["Io"] = "IO";
    /** Iraq */
    CountryCodeEnum["Iq"] = "IQ";
    /** Iran */
    CountryCodeEnum["Ir"] = "IR";
    /** Iceland */
    CountryCodeEnum["Is"] = "IS";
    /** Italy */
    CountryCodeEnum["It"] = "IT";
    /** Jersey */
    CountryCodeEnum["Je"] = "JE";
    /** Jamaica */
    CountryCodeEnum["Jm"] = "JM";
    /** Jordan */
    CountryCodeEnum["Jo"] = "JO";
    /** Japan */
    CountryCodeEnum["Jp"] = "JP";
    /** Kenya */
    CountryCodeEnum["Ke"] = "KE";
    /** Kyrgyzstan */
    CountryCodeEnum["Kg"] = "KG";
    /** Cambodia */
    CountryCodeEnum["Kh"] = "KH";
    /** Kiribati */
    CountryCodeEnum["Ki"] = "KI";
    /** Comoros */
    CountryCodeEnum["Km"] = "KM";
    /** St. Kitts & Nevis */
    CountryCodeEnum["Kn"] = "KN";
    /** North Korea */
    CountryCodeEnum["Kp"] = "KP";
    /** South Korea */
    CountryCodeEnum["Kr"] = "KR";
    /** Kuwait */
    CountryCodeEnum["Kw"] = "KW";
    /** Cayman Islands */
    CountryCodeEnum["Ky"] = "KY";
    /** Kazakhstan */
    CountryCodeEnum["Kz"] = "KZ";
    /** Laos */
    CountryCodeEnum["La"] = "LA";
    /** Lebanon */
    CountryCodeEnum["Lb"] = "LB";
    /** St. Lucia */
    CountryCodeEnum["Lc"] = "LC";
    /** Liechtenstein */
    CountryCodeEnum["Li"] = "LI";
    /** Sri Lanka */
    CountryCodeEnum["Lk"] = "LK";
    /** Liberia */
    CountryCodeEnum["Lr"] = "LR";
    /** Lesotho */
    CountryCodeEnum["Ls"] = "LS";
    /** Lithuania */
    CountryCodeEnum["Lt"] = "LT";
    /** Luxembourg */
    CountryCodeEnum["Lu"] = "LU";
    /** Latvia */
    CountryCodeEnum["Lv"] = "LV";
    /** Libya */
    CountryCodeEnum["Ly"] = "LY";
    /** Morocco */
    CountryCodeEnum["Ma"] = "MA";
    /** Monaco */
    CountryCodeEnum["Mc"] = "MC";
    /** Moldova */
    CountryCodeEnum["Md"] = "MD";
    /** Montenegro */
    CountryCodeEnum["Me"] = "ME";
    /** St. Martin */
    CountryCodeEnum["Mf"] = "MF";
    /** Madagascar */
    CountryCodeEnum["Mg"] = "MG";
    /** Marshall Islands */
    CountryCodeEnum["Mh"] = "MH";
    /** Macedonia */
    CountryCodeEnum["Mk"] = "MK";
    /** Mali */
    CountryCodeEnum["Ml"] = "ML";
    /** Myanmar (Burma) */
    CountryCodeEnum["Mm"] = "MM";
    /** Mongolia */
    CountryCodeEnum["Mn"] = "MN";
    /** Macau SAR China */
    CountryCodeEnum["Mo"] = "MO";
    /** Northern Mariana Islands */
    CountryCodeEnum["Mp"] = "MP";
    /** Martinique */
    CountryCodeEnum["Mq"] = "MQ";
    /** Mauritania */
    CountryCodeEnum["Mr"] = "MR";
    /** Montserrat */
    CountryCodeEnum["Ms"] = "MS";
    /** Malta */
    CountryCodeEnum["Mt"] = "MT";
    /** Mauritius */
    CountryCodeEnum["Mu"] = "MU";
    /** Maldives */
    CountryCodeEnum["Mv"] = "MV";
    /** Malawi */
    CountryCodeEnum["Mw"] = "MW";
    /** Mexico */
    CountryCodeEnum["Mx"] = "MX";
    /** Malaysia */
    CountryCodeEnum["My"] = "MY";
    /** Mozambique */
    CountryCodeEnum["Mz"] = "MZ";
    /** Namibia */
    CountryCodeEnum["Na"] = "NA";
    /** New Caledonia */
    CountryCodeEnum["Nc"] = "NC";
    /** Niger */
    CountryCodeEnum["Ne"] = "NE";
    /** Norfolk Island */
    CountryCodeEnum["Nf"] = "NF";
    /** Nigeria */
    CountryCodeEnum["Ng"] = "NG";
    /** Nicaragua */
    CountryCodeEnum["Ni"] = "NI";
    /** Netherlands */
    CountryCodeEnum["Nl"] = "NL";
    /** Norway */
    CountryCodeEnum["No"] = "NO";
    /** Nepal */
    CountryCodeEnum["Np"] = "NP";
    /** Nauru */
    CountryCodeEnum["Nr"] = "NR";
    /** Niue */
    CountryCodeEnum["Nu"] = "NU";
    /** New Zealand */
    CountryCodeEnum["Nz"] = "NZ";
    /** Oman */
    CountryCodeEnum["Om"] = "OM";
    /** Panama */
    CountryCodeEnum["Pa"] = "PA";
    /** Peru */
    CountryCodeEnum["Pe"] = "PE";
    /** French Polynesia */
    CountryCodeEnum["Pf"] = "PF";
    /** Papua New Guinea */
    CountryCodeEnum["Pg"] = "PG";
    /** Philippines */
    CountryCodeEnum["Ph"] = "PH";
    /** Pakistan */
    CountryCodeEnum["Pk"] = "PK";
    /** Poland */
    CountryCodeEnum["Pl"] = "PL";
    /** St. Pierre & Miquelon */
    CountryCodeEnum["Pm"] = "PM";
    /** Pitcairn Islands */
    CountryCodeEnum["Pn"] = "PN";
    /** Palestinian Territories */
    CountryCodeEnum["Ps"] = "PS";
    /** Portugal */
    CountryCodeEnum["Pt"] = "PT";
    /** Palau */
    CountryCodeEnum["Pw"] = "PW";
    /** Paraguay */
    CountryCodeEnum["Py"] = "PY";
    /** Qatar */
    CountryCodeEnum["Qa"] = "QA";
    /** Réunion */
    CountryCodeEnum["Re"] = "RE";
    /** Romania */
    CountryCodeEnum["Ro"] = "RO";
    /** Serbia */
    CountryCodeEnum["Rs"] = "RS";
    /** Russia */
    CountryCodeEnum["Ru"] = "RU";
    /** Rwanda */
    CountryCodeEnum["Rw"] = "RW";
    /** Saudi Arabia */
    CountryCodeEnum["Sa"] = "SA";
    /** Solomon Islands */
    CountryCodeEnum["Sb"] = "SB";
    /** Seychelles */
    CountryCodeEnum["Sc"] = "SC";
    /** Sudan */
    CountryCodeEnum["Sd"] = "SD";
    /** Sweden */
    CountryCodeEnum["Se"] = "SE";
    /** Singapore */
    CountryCodeEnum["Sg"] = "SG";
    /** St. Helena */
    CountryCodeEnum["Sh"] = "SH";
    /** Slovenia */
    CountryCodeEnum["Si"] = "SI";
    /** Svalbard & Jan Mayen */
    CountryCodeEnum["Sj"] = "SJ";
    /** Slovakia */
    CountryCodeEnum["Sk"] = "SK";
    /** Sierra Leone */
    CountryCodeEnum["Sl"] = "SL";
    /** San Marino */
    CountryCodeEnum["Sm"] = "SM";
    /** Senegal */
    CountryCodeEnum["Sn"] = "SN";
    /** Somalia */
    CountryCodeEnum["So"] = "SO";
    /** Suriname */
    CountryCodeEnum["Sr"] = "SR";
    /** São Tomé & Príncipe */
    CountryCodeEnum["St"] = "ST";
    /** El Salvador */
    CountryCodeEnum["Sv"] = "SV";
    /** Syria */
    CountryCodeEnum["Sy"] = "SY";
    /** Swaziland */
    CountryCodeEnum["Sz"] = "SZ";
    /** Turks & Caicos Islands */
    CountryCodeEnum["Tc"] = "TC";
    /** Chad */
    CountryCodeEnum["Td"] = "TD";
    /** French Southern Territories */
    CountryCodeEnum["Tf"] = "TF";
    /** Togo */
    CountryCodeEnum["Tg"] = "TG";
    /** Thailand */
    CountryCodeEnum["Th"] = "TH";
    /** Tajikistan */
    CountryCodeEnum["Tj"] = "TJ";
    /** Tokelau */
    CountryCodeEnum["Tk"] = "TK";
    /** Timor-Leste */
    CountryCodeEnum["Tl"] = "TL";
    /** Turkmenistan */
    CountryCodeEnum["Tm"] = "TM";
    /** Tunisia */
    CountryCodeEnum["Tn"] = "TN";
    /** Tonga */
    CountryCodeEnum["To"] = "TO";
    /** Turkey */
    CountryCodeEnum["Tr"] = "TR";
    /** Trinidad & Tobago */
    CountryCodeEnum["Tt"] = "TT";
    /** Tuvalu */
    CountryCodeEnum["Tv"] = "TV";
    /** Taiwan */
    CountryCodeEnum["Tw"] = "TW";
    /** Tanzania */
    CountryCodeEnum["Tz"] = "TZ";
    /** Ukraine */
    CountryCodeEnum["Ua"] = "UA";
    /** Uganda */
    CountryCodeEnum["Ug"] = "UG";
    /** U.S. Outlying Islands */
    CountryCodeEnum["Um"] = "UM";
    /** United States */
    CountryCodeEnum["Us"] = "US";
    /** Uruguay */
    CountryCodeEnum["Uy"] = "UY";
    /** Uzbekistan */
    CountryCodeEnum["Uz"] = "UZ";
    /** Vatican City */
    CountryCodeEnum["Va"] = "VA";
    /** St. Vincent & Grenadines */
    CountryCodeEnum["Vc"] = "VC";
    /** Venezuela */
    CountryCodeEnum["Ve"] = "VE";
    /** British Virgin Islands */
    CountryCodeEnum["Vg"] = "VG";
    /** U.S. Virgin Islands */
    CountryCodeEnum["Vi"] = "VI";
    /** Vietnam */
    CountryCodeEnum["Vn"] = "VN";
    /** Vanuatu */
    CountryCodeEnum["Vu"] = "VU";
    /** Wallis & Futuna */
    CountryCodeEnum["Wf"] = "WF";
    /** Samoa */
    CountryCodeEnum["Ws"] = "WS";
    /** Yemen */
    CountryCodeEnum["Ye"] = "YE";
    /** Mayotte */
    CountryCodeEnum["Yt"] = "YT";
    /** South Africa */
    CountryCodeEnum["Za"] = "ZA";
    /** Zambia */
    CountryCodeEnum["Zm"] = "ZM";
    /** Zimbabwe */
    CountryCodeEnum["Zw"] = "ZW";
})(exports.CountryCodeEnum || (exports.CountryCodeEnum = {}));
/** The list of available currency codes */
exports.CurrencyEnum = void 0;
(function (CurrencyEnum) {
    CurrencyEnum["Aed"] = "AED";
    CurrencyEnum["Afn"] = "AFN";
    CurrencyEnum["All"] = "ALL";
    CurrencyEnum["Amd"] = "AMD";
    CurrencyEnum["Ang"] = "ANG";
    CurrencyEnum["Aoa"] = "AOA";
    CurrencyEnum["Ars"] = "ARS";
    CurrencyEnum["Aud"] = "AUD";
    CurrencyEnum["Awg"] = "AWG";
    CurrencyEnum["Azm"] = "AZM";
    CurrencyEnum["Azn"] = "AZN";
    CurrencyEnum["Bam"] = "BAM";
    CurrencyEnum["Bbd"] = "BBD";
    CurrencyEnum["Bdt"] = "BDT";
    CurrencyEnum["Bgn"] = "BGN";
    CurrencyEnum["Bhd"] = "BHD";
    CurrencyEnum["Bif"] = "BIF";
    CurrencyEnum["Bmd"] = "BMD";
    CurrencyEnum["Bnd"] = "BND";
    CurrencyEnum["Bob"] = "BOB";
    CurrencyEnum["Brl"] = "BRL";
    CurrencyEnum["Bsd"] = "BSD";
    CurrencyEnum["Btn"] = "BTN";
    CurrencyEnum["Buk"] = "BUK";
    CurrencyEnum["Bwp"] = "BWP";
    CurrencyEnum["Byn"] = "BYN";
    CurrencyEnum["Bzd"] = "BZD";
    CurrencyEnum["Cad"] = "CAD";
    CurrencyEnum["Cdf"] = "CDF";
    CurrencyEnum["Che"] = "CHE";
    CurrencyEnum["Chf"] = "CHF";
    CurrencyEnum["Chw"] = "CHW";
    CurrencyEnum["Clp"] = "CLP";
    CurrencyEnum["Cny"] = "CNY";
    CurrencyEnum["Cop"] = "COP";
    CurrencyEnum["Crc"] = "CRC";
    CurrencyEnum["Cup"] = "CUP";
    CurrencyEnum["Cve"] = "CVE";
    CurrencyEnum["Czk"] = "CZK";
    CurrencyEnum["Djf"] = "DJF";
    CurrencyEnum["Dkk"] = "DKK";
    CurrencyEnum["Dop"] = "DOP";
    CurrencyEnum["Dzd"] = "DZD";
    CurrencyEnum["Eek"] = "EEK";
    CurrencyEnum["Egp"] = "EGP";
    CurrencyEnum["Ern"] = "ERN";
    CurrencyEnum["Etb"] = "ETB";
    CurrencyEnum["Eur"] = "EUR";
    CurrencyEnum["Fjd"] = "FJD";
    CurrencyEnum["Fkp"] = "FKP";
    CurrencyEnum["Gbp"] = "GBP";
    CurrencyEnum["Gek"] = "GEK";
    CurrencyEnum["Gel"] = "GEL";
    CurrencyEnum["Ghs"] = "GHS";
    CurrencyEnum["Gip"] = "GIP";
    CurrencyEnum["Gmd"] = "GMD";
    CurrencyEnum["Gnf"] = "GNF";
    CurrencyEnum["Gqe"] = "GQE";
    CurrencyEnum["Gtq"] = "GTQ";
    CurrencyEnum["Gyd"] = "GYD";
    CurrencyEnum["Hkd"] = "HKD";
    CurrencyEnum["Hnl"] = "HNL";
    CurrencyEnum["Hrk"] = "HRK";
    CurrencyEnum["Htg"] = "HTG";
    CurrencyEnum["Huf"] = "HUF";
    CurrencyEnum["Idr"] = "IDR";
    CurrencyEnum["Ils"] = "ILS";
    CurrencyEnum["Inr"] = "INR";
    CurrencyEnum["Iqd"] = "IQD";
    CurrencyEnum["Irr"] = "IRR";
    CurrencyEnum["Isk"] = "ISK";
    CurrencyEnum["Jmd"] = "JMD";
    CurrencyEnum["Jod"] = "JOD";
    CurrencyEnum["Jpy"] = "JPY";
    CurrencyEnum["Kes"] = "KES";
    CurrencyEnum["Kgs"] = "KGS";
    CurrencyEnum["Khr"] = "KHR";
    CurrencyEnum["Kmf"] = "KMF";
    CurrencyEnum["Kpw"] = "KPW";
    CurrencyEnum["Krw"] = "KRW";
    CurrencyEnum["Kwd"] = "KWD";
    CurrencyEnum["Kyd"] = "KYD";
    CurrencyEnum["Kzt"] = "KZT";
    CurrencyEnum["Lak"] = "LAK";
    CurrencyEnum["Lbp"] = "LBP";
    CurrencyEnum["Lkr"] = "LKR";
    CurrencyEnum["Lrd"] = "LRD";
    CurrencyEnum["Lsl"] = "LSL";
    CurrencyEnum["Lsm"] = "LSM";
    CurrencyEnum["Ltl"] = "LTL";
    CurrencyEnum["Lvl"] = "LVL";
    CurrencyEnum["Lyd"] = "LYD";
    CurrencyEnum["Mad"] = "MAD";
    CurrencyEnum["Mdl"] = "MDL";
    CurrencyEnum["Mga"] = "MGA";
    CurrencyEnum["Mkd"] = "MKD";
    CurrencyEnum["Mmk"] = "MMK";
    CurrencyEnum["Mnt"] = "MNT";
    CurrencyEnum["Mop"] = "MOP";
    CurrencyEnum["Mro"] = "MRO";
    CurrencyEnum["Mur"] = "MUR";
    CurrencyEnum["Mvr"] = "MVR";
    CurrencyEnum["Mwk"] = "MWK";
    CurrencyEnum["Mxn"] = "MXN";
    CurrencyEnum["Myr"] = "MYR";
    CurrencyEnum["Mzn"] = "MZN";
    CurrencyEnum["Nad"] = "NAD";
    CurrencyEnum["Ngn"] = "NGN";
    CurrencyEnum["Nic"] = "NIC";
    CurrencyEnum["Nok"] = "NOK";
    CurrencyEnum["Npr"] = "NPR";
    CurrencyEnum["Nzd"] = "NZD";
    CurrencyEnum["Omr"] = "OMR";
    CurrencyEnum["Pab"] = "PAB";
    CurrencyEnum["Pen"] = "PEN";
    CurrencyEnum["Pgk"] = "PGK";
    CurrencyEnum["Php"] = "PHP";
    CurrencyEnum["Pkr"] = "PKR";
    CurrencyEnum["Pln"] = "PLN";
    CurrencyEnum["Pyg"] = "PYG";
    CurrencyEnum["Qar"] = "QAR";
    CurrencyEnum["Rhd"] = "RHD";
    CurrencyEnum["Rol"] = "ROL";
    CurrencyEnum["Ron"] = "RON";
    CurrencyEnum["Rsd"] = "RSD";
    CurrencyEnum["Rub"] = "RUB";
    CurrencyEnum["Rwf"] = "RWF";
    CurrencyEnum["Sar"] = "SAR";
    CurrencyEnum["Sbd"] = "SBD";
    CurrencyEnum["Scr"] = "SCR";
    CurrencyEnum["Sdg"] = "SDG";
    CurrencyEnum["Sek"] = "SEK";
    CurrencyEnum["Sgd"] = "SGD";
    CurrencyEnum["Shp"] = "SHP";
    CurrencyEnum["Skk"] = "SKK";
    CurrencyEnum["Sll"] = "SLL";
    CurrencyEnum["Sos"] = "SOS";
    CurrencyEnum["Srd"] = "SRD";
    CurrencyEnum["Std"] = "STD";
    CurrencyEnum["Svc"] = "SVC";
    CurrencyEnum["Syp"] = "SYP";
    CurrencyEnum["Szl"] = "SZL";
    CurrencyEnum["Thb"] = "THB";
    CurrencyEnum["Tjs"] = "TJS";
    CurrencyEnum["Tmm"] = "TMM";
    CurrencyEnum["Tnd"] = "TND";
    CurrencyEnum["Top"] = "TOP";
    CurrencyEnum["Trl"] = "TRL";
    CurrencyEnum["Try"] = "TRY";
    CurrencyEnum["Ttd"] = "TTD";
    CurrencyEnum["Twd"] = "TWD";
    CurrencyEnum["Tzs"] = "TZS";
    CurrencyEnum["Uah"] = "UAH";
    CurrencyEnum["Ugx"] = "UGX";
    CurrencyEnum["Usd"] = "USD";
    CurrencyEnum["Uyu"] = "UYU";
    CurrencyEnum["Uzs"] = "UZS";
    CurrencyEnum["Veb"] = "VEB";
    CurrencyEnum["Vef"] = "VEF";
    CurrencyEnum["Vnd"] = "VND";
    CurrencyEnum["Vuv"] = "VUV";
    CurrencyEnum["Wst"] = "WST";
    CurrencyEnum["Xcd"] = "XCD";
    CurrencyEnum["Xof"] = "XOF";
    CurrencyEnum["Xpf"] = "XPF";
    CurrencyEnum["Yer"] = "YER";
    CurrencyEnum["Ytl"] = "YTL";
    CurrencyEnum["Zar"] = "ZAR";
    CurrencyEnum["Zmk"] = "ZMK";
    CurrencyEnum["Zwd"] = "ZWD";
})(exports.CurrencyEnum || (exports.CurrencyEnum = {}));
/** This enumeration customizable date type. */
exports.CustomizableDateTypeEnum = void 0;
(function (CustomizableDateTypeEnum) {
    CustomizableDateTypeEnum["Date"] = "DATE";
    CustomizableDateTypeEnum["DateTime"] = "DATE_TIME";
    CustomizableDateTypeEnum["Time"] = "TIME";
})(exports.CustomizableDateTypeEnum || (exports.CustomizableDateTypeEnum = {}));
exports.DownloadableFileTypeEnum = void 0;
(function (DownloadableFileTypeEnum) {
    DownloadableFileTypeEnum["File"] = "FILE";
    DownloadableFileTypeEnum["Url"] = "URL";
})(exports.DownloadableFileTypeEnum || (exports.DownloadableFileTypeEnum = {}));
/** This enumeration display settings for the fixed product tax */
exports.FixedProductTaxDisplaySettings = void 0;
(function (FixedProductTaxDisplaySettings) {
    /** The displayed price does not include the FPT amount. The values of ProductPrice.fixed_product_taxes and the price including the FPT are displayed separately. This value corresponds to 'Excluding FPT, Including FPT description and final price' */
    FixedProductTaxDisplaySettings["ExcludeFptAndIncludeWithDetails"] = "EXCLUDE_FPT_AND_INCLUDE_WITH_DETAILS";
    /** The displayed price does not include the FPT amount. The values from ProductPrice.fixed_product_taxes are not displayed. This value corresponds to 'Excluding FPT' */
    FixedProductTaxDisplaySettings["ExcludeFptWithoutDetails"] = "EXCLUDE_FPT_WITHOUT_DETAILS";
    /** The FPT feature is not enabled. You can omit  ProductPrice.fixed_product_taxes from your query */
    FixedProductTaxDisplaySettings["FptDisabled"] = "FPT_DISABLED";
    /** The displayed price includes the FPT amount without displaying the ProductPrice.fixed_product_taxes values. This value corresponds to 'Including FPT only' */
    FixedProductTaxDisplaySettings["IncludeFptWithoutDetails"] = "INCLUDE_FPT_WITHOUT_DETAILS";
    /** The displayed price includes the FPT amount while displaying the values of ProductPrice.fixed_product_taxes separately. This value corresponds to 'Including FPT and FPT description' */
    FixedProductTaxDisplaySettings["IncludeFptWithDetails"] = "INCLUDE_FPT_WITH_DETAILS";
})(exports.FixedProductTaxDisplaySettings || (exports.FixedProductTaxDisplaySettings = {}));
/** Mode for payment: TEST or LIVE. Applies to Payflow Link and Payments Advanced payment methods. */
exports.PayflowLinkMode = void 0;
(function (PayflowLinkMode) {
    PayflowLinkMode["Live"] = "LIVE";
    PayflowLinkMode["Test"] = "TEST";
})(exports.PayflowLinkMode || (exports.PayflowLinkMode = {}));
/** The list of available payment token types */
exports.PaymentTokenTypeEnum = void 0;
(function (PaymentTokenTypeEnum) {
    PaymentTokenTypeEnum["Account"] = "account";
    PaymentTokenTypeEnum["Card"] = "card";
})(exports.PaymentTokenTypeEnum || (exports.PaymentTokenTypeEnum = {}));
/** PriceAdjustment.code is deprecated. This enumeration contains values defined in modules other than the Catalog module. */
exports.PriceAdjustmentCodesEnum = void 0;
(function (PriceAdjustmentCodesEnum) {
    PriceAdjustmentCodesEnum["Tax"] = "TAX";
    PriceAdjustmentCodesEnum["Weee"] = "WEEE";
    PriceAdjustmentCodesEnum["WeeeTax"] = "WEEE_TAX";
})(exports.PriceAdjustmentCodesEnum || (exports.PriceAdjustmentCodesEnum = {}));
/** PriceAdjustmentDescriptionEnum is deprecated. This enumeration states whether a price adjustment is included or excluded. */
exports.PriceAdjustmentDescriptionEnum = void 0;
(function (PriceAdjustmentDescriptionEnum) {
    PriceAdjustmentDescriptionEnum["Excluded"] = "EXCLUDED";
    PriceAdjustmentDescriptionEnum["Included"] = "INCLUDED";
})(exports.PriceAdjustmentDescriptionEnum || (exports.PriceAdjustmentDescriptionEnum = {}));
/** This enumeration the price type. */
exports.PriceTypeEnum = void 0;
(function (PriceTypeEnum) {
    PriceTypeEnum["Dynamic"] = "DYNAMIC";
    PriceTypeEnum["Fixed"] = "FIXED";
    PriceTypeEnum["Percent"] = "PERCENT";
})(exports.PriceTypeEnum || (exports.PriceTypeEnum = {}));
/** This enumeration defines whether a bundle product's price is displayed as the lowest possible value or as a range. */
exports.PriceViewEnum = void 0;
(function (PriceViewEnum) {
    PriceViewEnum["AsLowAs"] = "AS_LOW_AS";
    PriceViewEnum["PriceRange"] = "PRICE_RANGE";
})(exports.PriceViewEnum || (exports.PriceViewEnum = {}));
/** This enumeration states whether a product stock status is in stock or out of stock */
exports.ProductStockStatus = void 0;
(function (ProductStockStatus) {
    ProductStockStatus["InStock"] = "IN_STOCK";
    ProductStockStatus["OutOfStock"] = "OUT_OF_STOCK";
})(exports.ProductStockStatus || (exports.ProductStockStatus = {}));
/** This enumeration defines whether bundle items must be shipped together. */
exports.ShipBundleItemsEnum = void 0;
(function (ShipBundleItemsEnum) {
    ShipBundleItemsEnum["Separately"] = "SEPARATELY";
    ShipBundleItemsEnum["Together"] = "TOGETHER";
})(exports.ShipBundleItemsEnum || (exports.ShipBundleItemsEnum = {}));
/** This enumeration indicates whether to return results in ascending or descending order */
exports.SortEnum = void 0;
(function (SortEnum) {
    SortEnum["Asc"] = "ASC";
    SortEnum["Desc"] = "DESC";
})(exports.SortEnum || (exports.SortEnum = {}));
exports.SubscriptionStatusesEnum = void 0;
(function (SubscriptionStatusesEnum) {
    SubscriptionStatusesEnum["NotActive"] = "NOT_ACTIVE";
    SubscriptionStatusesEnum["Subscribed"] = "SUBSCRIBED";
    SubscriptionStatusesEnum["Unconfirmed"] = "UNCONFIRMED";
    SubscriptionStatusesEnum["Unsubscribed"] = "UNSUBSCRIBED";
})(exports.SubscriptionStatusesEnum || (exports.SubscriptionStatusesEnum = {}));
/** This enumeration defines the entity type. */
exports.UrlRewriteEntityTypeEnum = void 0;
(function (UrlRewriteEntityTypeEnum) {
    UrlRewriteEntityTypeEnum["Category"] = "CATEGORY";
    UrlRewriteEntityTypeEnum["CmsPage"] = "CMS_PAGE";
    UrlRewriteEntityTypeEnum["Product"] = "PRODUCT";
})(exports.UrlRewriteEntityTypeEnum || (exports.UrlRewriteEntityTypeEnum = {}));
exports.UseInLayeredNavigationOptions = void 0;
(function (UseInLayeredNavigationOptions) {
    UseInLayeredNavigationOptions["FilterableNoResult"] = "FILTERABLE_NO_RESULT";
    UseInLayeredNavigationOptions["FilterableWithResults"] = "FILTERABLE_WITH_RESULTS";
    UseInLayeredNavigationOptions["No"] = "NO";
})(exports.UseInLayeredNavigationOptions || (exports.UseInLayeredNavigationOptions = {}));
exports.WishListUserInputErrorType = void 0;
(function (WishListUserInputErrorType) {
    WishListUserInputErrorType["ProductNotFound"] = "PRODUCT_NOT_FOUND";
    WishListUserInputErrorType["Undefined"] = "UNDEFINED";
})(exports.WishListUserInputErrorType || (exports.WishListUserInputErrorType = {}));
exports.WishlistCartUserInputErrorType = void 0;
(function (WishlistCartUserInputErrorType) {
    WishlistCartUserInputErrorType["InsufficientStock"] = "INSUFFICIENT_STOCK";
    WishlistCartUserInputErrorType["NotSalable"] = "NOT_SALABLE";
    WishlistCartUserInputErrorType["ProductNotFound"] = "PRODUCT_NOT_FOUND";
    WishlistCartUserInputErrorType["Undefined"] = "UNDEFINED";
})(exports.WishlistCartUserInputErrorType || (exports.WishlistCartUserInputErrorType = {}));

exports.MagentoCustomerGender = void 0;
(function (MagentoCustomerGender) {
    MagentoCustomerGender[MagentoCustomerGender["Male"] = 1] = "Male";
    MagentoCustomerGender[MagentoCustomerGender["Female"] = 2] = "Female";
})(exports.MagentoCustomerGender || (exports.MagentoCustomerGender = {}));
